security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -227,23 +231,73 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
                 }
             }
 
-            if (authHeader != null && !request.getServletPath().equals("/admin")) {
+            // if no OAuth token received check if credentials are provided in authorization header
+            // and use back channel direct auth flow
+            if (accessToken == null && authHeader != null && !request.getServletPath().equals("/admin")) {
+                try {
+                    var credentials = parseBasicAuth(authHeader);
+                    if (credentials == null) {
+                        var clientCredentialsToken = parseBearerToken(authHeader);
+
+                        if (clientCredentialsToken != null) {
+                            // if client credentials token, assume it's a JWT token that contains needed info
+                            log.debug("Client Credentials Token = {}", clientCredentialsToken);
+
+                            String[] chunks = clientCredentialsToken.split("\\.");
+
+                            Base64.Decoder decoder = Base64.getUrlDecoder();
+
+                            String header = new String(decoder.decode(chunks[0]));
+
+                            log.debug("Client Credentials Token Header = {}", header);
+
+                            JsonElement headerData = JsonParser.parseString(header);
 
-                clientCredentialsToken = parseBearerToken(authHeader);
+                            JwkProvider jwkProvider =
+                                    new JwkProviderBuilder(new URL(config.bearerTokenConfig.jwksUri))
+                                            .cached(config.bearerTokenConfig.cacheSize,
+                                                    Duration.of(config.bearerTokenConfig.cacheDuration, ChronoUnit.MINUTES))
+                                            .build();
```

Comment:
I think to benefit from the cache this would have to be created elsewhere, right?
Here it looks like it's recreated on every auth request

Reply:
In this case, since a cache size and duration are specified the JwkProvider invokes a GuavaCachedJwkProvider which maintains an internal cache for the number of specified elements and their duration. 

https://javadoc.io/static/com.auth0/jwks-rsa/0.22.1/com/auth0/jwk/JwkProviderBuilder.html

https://javadoc.io/doc/com.auth0/jwks-rsa/0.3.0/com/auth0/jwk/GuavaCachedJwkProvider.html

![image](https://github.com/user-attachments/assets/cfe1cec4-2cd0-4b92-8dc7-278312dc8a87)

Reworked logic so that there is only ever one JwksProvider that is created on construction of authenticator or reset (recreated) when expired.